File: packages/core/src/gitmsg/parser.ts
/**
 * GitMsg protocol parser
 */

import { type GitMsgHeader, type GitMsgMessage, type GitMsgRef, isMessageType } from './types';

/**
 * Parse a GitMsg header line into structured data
 */
export function parseGitMsgHeader(headerLine: string): GitMsgHeader | null {
  // Header format: --- GitMsg: ext="..."; field="..."; v="..."; ext-v="..." ---
  const match = headerLine.match(/^--- GitMsg: (.*) ---$/);
  if (!match) {
    return null;
  }

  const fields: Record<string, string> = {};
  let ext = '';
  let v = '';
  let extV = '';

  // Parse semicolon-separated key="value" pairs
  const fieldRegex = /([a-zA-Z_][a-zA-Z0-9_:-]*)="([^"]*)"/g;
  let fieldMatch;
  const headerContent = match[1] || '';

  while ((fieldMatch = fieldRegex.exec(headerContent)) !== null) {
    const [, key, value] = fieldMatch;

    if (key && value !== undefined) {
      switch (key) {
      case 'ext':
        ext = value;
        break;
      case 'v':
        v = value;
        break;
      case 'ext-v':
        extV = value;
        break;
      default:
        // Extension-specific fields (e.g., "type", "original")
        fields[key] = value;
      }
    }
  }

  // Validate required fields
  if (!ext || !v || !extV) {
    return null;
  }

  return {
    ext,
    v,
    extV,
    fields
  };
}

/**
 * Parse a GitMsg reference section
 */
export function parseGitMsgRef(refSection: string): GitMsgRef | null {
  const lines = refSection.split('\n');
  const headerLine = lines[0];

  if (!headerLine) {
    return null;
  }

  // Reference format: --- GitMsg-Ref: ext="..."; author="..."; ref="..."; v="..."; ext-v="..." ---
  const match = headerLine.match(/^--- GitMsg-Ref: (.*) ---$/);
  if (!match) {
    return null;
  }

  const fields: Record<string, string> = {};
  let ext = '';
  let ref = '';
  let v = '';
  let extV = '';
  let author: string | undefined;
  let email: string | undefined;
  let time: string | undefined;

  // Parse fields from header line
  const fieldRegex = /([a-zA-Z_][a-zA-Z0-9_:-]*)="([^"]*)"/g;
  let fieldMatch;
  const refContent = match[1] || '';

  while ((fieldMatch = fieldRegex.exec(refContent)) !== null) {
    const [, key, value] = fieldMatch;

    if (key && value !== undefined) {
      switch (key) {
      case 'ext':
        ext = value;
        break;
      case 'ref':
        ref = value;
        break;
      case 'v':
        v = value;
        break;
      case 'ext-v':
        extV = value;
        break;
      case 'author':
        author = value;
        break;
      case 'email':
        email = value;
        break;
      case 'time':
        time = value;
        break;
      default:
        // Extension-specific fields
        fields[key] = value;
      }
    }
  }

  // Validate required fields
  if (!ext || !ref || !v || !extV || !author || !email || !time) {
    return null;
  }

  // Get metadata content (everything after the header line)
  const metadata = lines.slice(1).join('\n').trim() || undefined;

  return {
    ext,
    ref,
    v,
    extV,
    author,
    email,
    time,
    fields,
    metadata
  };
}

/**
 * Parse a complete GitMsg message
 */
export function parseGitMsgMessage(message: string): GitMsgMessage | null {
  // Find the GitMsg header
  const headerMatch = message.match(/^([\s\S]*?)\n--- GitMsg: .* ---$/m);
  if (!headerMatch || !headerMatch[1]) {
    return null;
  }

  const content = headerMatch[1].trim();
  const headerStart = headerMatch.index! + headerMatch[1].length + 1;

  // Extract header line
  const headerEnd = message.indexOf('---', headerStart + 11) + 3;
  const headerLine = message.substring(headerStart, headerEnd);

  const header = parseGitMsgHeader(headerLine);
  if (!header) {
    return null;
  }

  // Parse reference sections
  const references: GitMsgRef[] = [];
  const refRegex = /--- GitMsg-Ref: .*? ---[\s\S]*?(?=--- GitMsg-Ref:|$)/g;
  const remainingMessage = message.substring(headerEnd);

  let refMatch;
  while ((refMatch = refRegex.exec(remainingMessage)) !== null) {
    const ref = parseGitMsgRef(refMatch[0].trim());
    if (ref) {
      references.push(ref);
    }
  }

  return {
    content,
    header,
    references
  };
}

/**
 * Validate a GitMsg message structure
 */
export function validateGitMsgMessage(message: GitMsgMessage | null): boolean {
  if (!message) {
    return false;
  }

  // Check header
  if (!message.header.ext || !message.header.v || !message.header.extV) {
    return false;
  }

  // Validate extension format
  if (!message.header.ext.match(/^[a-z][a-z0-9_-]*$/)) {
    return false;
  }

  // Validate version format
  const versionRegex = /^\d+\.\d+\.\d+$/;
  if (!message.header.v.match(versionRegex) ||
      !message.header.extV.match(versionRegex)) {
    return false;
  }

  // Validate references
  for (const ref of message.references) {
    if (!ref.ext.match(/^[a-z][a-z0-9_-]*$/)) {
      return false;
    }

    // Validate ref format - check for valid reference patterns
    // Commits must have exactly 12-character hashes, branches can be any valid name
    const refPattern = new RegExp(
      '^((https?:\\/\\/[^#\\s]+|[^#\\s]+)#(commit:[a-f0-9]{12}|branch:[a-zA-Z0-9/_-]+)|' +
      '#(commit:[a-f0-9]{12}|branch:[a-zA-Z0-9/_-]+))$'
    );
    if (!refPattern.test(ref.ref)) {
      return false;
    }

    if (!ref.v.match(versionRegex) || !ref.extV.match(versionRegex)) {
      return false;
    }
  }

  return true;
}

/**
 * Extract clean content from commit message (remove GitMsg headers)
 * Pure utility function for text processing
 */
export function extractCleanContent(message: string): string {
  // Remove GitMsg header
  const headerRegex = /^--- GitMsg:.*?---$/m;
  let cleanMessage = message.replace(headerRegex, '').trim();

  // Remove GitMsg-Ref sections
  const refRegex = /^--- GitMsg-Ref:.*?---[\s\S]*?(?=^--- GitMsg-Ref:|$(?!\n))/gm;
  cleanMessage = cleanMessage.replace(refRegex, '').trim();

  return cleanMessage;
}

/**
 * Determine post type from GitMsg header
 * Pure function for type detection
 */
export function getPostType(gitMsg?: GitMsgMessage): 'post' | 'comment' | 'repost' | 'quote' {
  if (!gitMsg) {return 'post';}
  if (gitMsg.header.ext !== 'social') {return 'post';}

  const socialType = gitMsg.header.fields['type'];
  switch (socialType) {
  case 'comment':
    return 'comment';
  case 'repost':
    return 'repost';
  case 'quote':
    return 'quote';
  case 'post':
  default:
    return 'post';
  }
}

/**
 * Check if a repost is empty (simple repost without additional content)
 * Pure predicate function
 */
export function isEmptyRepost(gitMsg: GitMsgMessage): boolean {
  if (!isMessageType(gitMsg.header, 'social', 'repost')) {return false;}

  // A repost is empty if it only contains the attribution line (starts with #)
  const content = gitMsg.content.trim();
  return content.startsWith('#') && content.split('\n').length === 1;
}

File: library/core/protocol/parser.go
// parser.go - GitMsg protocol message parsing
package protocol

import (
	"regexp"
	"strings"
)

var (
	headerLinePattern = regexp.MustCompile(`^GitMsg: (.*)$`)
	refLinePattern    = regexp.MustCompile(`^GitMsg-Ref: (.*)$`)
	fieldPattern      = regexp.MustCompile(`([a-zA-Z_][a-zA-Z0-9_:-]*)="([^"]*)"`)
)

// ParseHeader parses a GitMsg trailer line into a Header struct.
func ParseHeader(headerLine string) *Header {
	match := headerLinePattern.FindStringSubmatch(headerLine)
	if len(match) < 2 {
		return nil
	}

	fields := make(map[string]string)
	var ext, v string

	headerContent := match[1]
	fieldMatches := fieldPattern.FindAllStringSubmatch(headerContent, -1)

	for _, fm := range fieldMatches {
		if len(fm) < 3 {
			continue
		}
		key, value := fm[1], fm[2]
		switch key {
		case "ext":
			ext = value
		case "v":
			v = value
		default:
			fields[key] = value
		}
	}

	if ext == "" || v == "" {
		return nil
	}

	return &Header{
		Ext:    ext,
		V:      v,
		Fields: fields,
	}
}

// ParseRefSection parses a GitMsg-Ref trailer with continuation lines into a Ref struct.
func ParseRefSection(refSection string) *Ref {
	lines := strings.Split(refSection, "\n")
	if len(lines) == 0 {
		return nil
	}

	headerLine := lines[0]
	match := refLinePattern.FindStringSubmatch(headerLine)
	if len(match) < 2 {
		return nil
	}

	fields := make(map[string]string)
	var ext, ref, v, author, email, time string

	refContent := match[1]
	fieldMatches := fieldPattern.FindAllStringSubmatch(refContent, -1)

	for _, fm := range fieldMatches {
		if len(fm) < 3 {
			continue
		}
		key, value := fm[1], fm[2]
		switch key {
		case "ext":
			ext = value
		case "ref":
			ref = value
		case "v":
			v = value
		case "author":
			author = value
		case "email":
			email = value
		case "time":
			time = value
		default:
			fields[key] = value
		}
	}

	if ext == "" || ref == "" || v == "" || author == "" || email == "" || time == "" {
		return nil
	}

	metadata := ""
	if len(lines) > 1 {
		metaLines := make([]string, 0, len(lines)-1)
		for _, line := range lines[1:] {
			// Strip leading space (trailer continuation marker)
			if strings.HasPrefix(line, " ") {
				metaLines = append(metaLines, line[1:])
			} else {
				metaLines = append(metaLines, line)
			}
		}
		metadata = strings.TrimSpace(strings.Join(metaLines, "\n"))
	}

	return &Ref{
		Ext:      ext,
		Ref:      ref,
		V:        v,
		Author:   author,
		Email:    email,
		Time:     time,
		Fields:   fields,
		Metadata: metadata,
	}
}

// ParseMessage parses a complete GitMsg message with header and refs.
func ParseMessage(message string) *Message {
	// Find the GitMsg trailer (preceded by newline or at start of message)
	headerIdx := -1
	if strings.HasPrefix(message, "GitMsg: ") {
		headerIdx = 0
	} else {
		idx := strings.Index(message, "\nGitMsg: ")
		if idx != -1 {
			headerIdx = idx + 1
		}
	}
	if headerIdx == -1 {
		return nil
	}

	content := ""
	if headerIdx > 0 {
		content = strings.TrimSpace(message[:headerIdx])
	}

	// Split trailer block into lines
	trailerBlock := message[headerIdx:]
	lines := strings.Split(trailerBlock, "\n")

	// First line is the GitMsg header
	header := ParseHeader(lines[0])
	if header == nil {
		return nil
	}

	// Parse remaining lines for GitMsg-Ref trailers with continuation lines
	var references []Ref
	var currentRefLines []string

	for _, line := range lines[1:] {
		if strings.HasPrefix(line, "GitMsg-Ref: ") {
			// Flush previous ref if any
			if len(currentRefLines) > 0 {
				ref := ParseRefSection(strings.Join(currentRefLines, "\n"))
				if ref != nil {
					references = append(references, *ref)
				}
			}
			currentRefLines = []string{line}
		} else if strings.HasPrefix(line, " ") && len(currentRefLines) > 0 {
			// Continuation line for current ref
			currentRefLines = append(currentRefLines, line)
		}
	}
	// Flush last ref
	if len(currentRefLines) > 0 {
		ref := ParseRefSection(strings.Join(currentRefLines, "\n"))
		if ref != nil {
			references = append(references, *ref)
		}
	}

	return &Message{
		Content:    content,
		Header:     *header,
		References: references,
	}
}

// ExtractCleanContent extracts user content without protocol metadata.
func ExtractCleanContent(message string) string {
	headerIdx := -1
	if strings.HasPrefix(message, "GitMsg: ") {
		headerIdx = 0
	} else {
		idx := strings.Index(message, "\nGitMsg: ")
		if idx != -1 {
			headerIdx = idx
		}
	}
	content := message
	if headerIdx != -1 {
		content = message[:headerIdx]
	}
	content = strings.ReplaceAll(content, "\r", "")
	return strings.TrimSpace(content)
}
